### PR TITLE
Fix bsp: getMillBuildClasspath return URI string

### PR DESCRIPTION
**Pull request by camper42**
getMillBuildClasspath is expect to return URI, otherwise IDEA bsp import will have
`java.lang.IllegalArgumentException: Missing scheme` error on MacOS and
`Java.net.URISyntaxException: Illegal character in opaque part at index 2: ...` on Windows

### CODE REVIEW

**Comment by camper42** [comment]
problem I mentioned at https://github.com/com-lihaoyi/mill/pull/1285#issuecomment-842305505
seems caused by https://github.com/com-lihaoyi/mill/pull/1285/files#diff-a8dcb229dc6fe3ad91d4667a94174520d2d13cf522738fed08023105ac955d00R165

`Path.wrapped` return `java.nio.file.Path` and `toUri` is missing

@lefou @ScalaWilliam pls review

File: bsp/src/mill/bsp/ModuleUtils.scala [hunk 1]
```diff
@@ -162,7 +162,7 @@ object ModuleUtils {
     val binarySource =
       if (sources) all.filter(url => isPathSourceJar(url))
       else all.filter(url => !isPathSourceJar(url))
-    binarySource.filter(path => exists(path)).map(_.wrapped.toString)
+    binarySource.filter(path => exists(path)).map(_.toNIO.toUri.toString)
```
Comment (lefou) [inline bsp/src/mill/bsp/ModuleUtils.scala:165]: I'm wondering why we don't use a better return type here. At least, we should add some Scaladoc to line 126 to make it clear, that the returned `String`s are meant to be URIs.

**Review by lefou — COMMENTED** [review]
As it seems this code isn't covered by any tests, I'd like to have at least one written "It works for me" from someone who really tested this locally, for each platform.

**Comment by camper42** [comment]
> As it seems this code isn't covered by any tests, I'd like to have at least one written "It works for me" from someone who really tested this locally, for each platform.

will try to add an ut for at least getMillBuildClasspath first, If it seems to take a long time to do, I'm happy to write "It works for me" to get this fix merged.

**Comment by ScalaWilliam** [comment]
My apologies, I did not get a chance to test this.

**Comment by camper42** [comment]
hmmm，getMillBuildClasspath on `TestEvaluator.static(TestModule).evaluator` return empty list
so test always pass....

**Comment by camper42** [comment]
> getMillBuildClasspath on TestEvaluator.static(TestModule).evaluator return empty list

test is running on a version like `0.9.7-4-c85970-DIRTY55902d6c`, which can't access before `ci/publish-local.sh`, so:
* `classpath` in `getMillBuildClasspath` is empty due to `Try(evaluator.rootModule.getClass.getClassLoader.asInstanceOf[SpecialClassLoader])` failed
* `millJars` in `getMillBuildClasspath` is empty due to  `Resolution failed`

It's impossiable to test, choose to leave a "It works for me on MacOS IDEA" comment...

**Comment by camper42** [comment]
also worked for my Windows IDEA

`test-windows (8, cmd /C %GITHUB_WORKSPACE%\ci\mill.bat -i -d -k "{main,scalajslib,bsp}.__.test")` failure can be ignored?
nothing change about `mill.scalajslib`

**Review by lefou — COMMENTED** [review]
The CI failure is some flakiness in GitHub Actions, I reran the tests, which are now green.

I think the test issue(s) can be solved by propagating the freshly built jars into the test setup via `forkEnv` and/or `forkArgs`. We do this in other integration tests and it works. But I think it's too much to ask for in a bug fix PR and your "works for me" is better than what we have currently. We have lots of open building sites in our BSP story, and integration testing is a central part of it. 

Your technical change looks good to me, but the Scaladoc needs to be moved before the start of the `def` on line 126, not inside the `def`. Otherwise, we're good to merge.

**Comment by fabianhjr** [comment]
Had same issue on Linux (NixOS), ( `java.lang.IllegalArgumentException: Missing scheme`)

Tested this PR by:

1. Checking out this repo
2. Checking out 0.9.7
3. Cherrypicking this change
4. Retagging 0.9.7 to include this change
5. hand modifying a local project  `.bsp/mill.json` to point to a symbolic link to the run script (`/home/fabian/.local/bin/mill-checkout -> /home/fabian/Development/Open/mill/out/dev/launcher/dest/run`)
6. Restarting IntelliJ IDEA

This did fix the error and I was able to import the project again; local retagging was to avoid attempting to fetch nonexistant versions of mill-contrib-{bloop, scoverage} (0.9.7-[n]-[local commit that isn't available elsewhere]

**Comment by camper42** [comment]
@fabianhjr would you mind if I add "works for @fabianhjr on Linux(NixOS)" ?

**Comment by fabianhjr** [comment]
Not at all, go ahead.

Many thanks for the diagnosis and fix of this issue! UwU

**Comment by lefou** [comment]
Thanks you!

**Comment by sequencer** [comment]
When can we get this patch in a new version like 0.9.8? Since 0.9.7 of mill+bsp+IDEA is totally broken since this issue.

**Comment by lefou** [comment]
You can start to use it right away. Just pick the latest snapshot-build version from https://github.com/com-lihaoyi/mill/releases (e.g. `0.9.7-13-bf96a1`) and put it into `.mill-version` in your project top directory. See https://com-lihaoyi.github.io/mill/mill/Intro_to_Mill.html#_overriding_mill_versions

**Comment by lefou** [comment]
I'd like to get some other fixes into main branch before the next tag.

**Comment by sequencer** [comment]
Got it! Thank you!
I'll use `0.9.7-13-bf96a1` fix this temporarily, since I'm using mill package from Arch Linux.

**Comment by sequencer** [comment]
opps, seems that package from Arch Linux doesn't honor `0.9.7-13-bf96a1`

**Pull request merged**